packages/kinvey-datastore/src/repositories/network-repository.js

```diff
@@ -0,0 +1,99 @@
+import { KinveyRequest, RequestMethod, DeltaFetchRequest } from 'kinvey-request';
+
+import { Repository } from './repository';
+import { buildCollectionUrl, ensureArray } from '../utils';
+
+/**
+ * @typedef RequestOptions
+ * @property {Operation} operation
+ * @property {Object} options
+ * @property {Query} [options.query]
+ * @property {Query} [options.query]
+ * @property {Object} [options.aggregation]
+ * @property {Number} [options.timeout]
+ * @property {Object} [options.properties]
+ * @property {Boolean} [options.skipBL]
+ * @property {Boolean} [options.trace]
+ */
+
+export class NetworkRepository extends Repository {
+  read(collection, query, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.GET, null, query, null, options = {});
+    return this._makeHttpRequest(requestConfig, options.useDeltaFetch);
+  }
+
+  readById(collection, entityId, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.GET, null, null, entityId, options);
+    return this._makeHttpRequest(requestConfig);
+  }
+
+  create(collection, entities, options) {
+    return this._processBatch(collection, RequestMethod.POST, entities, null, options);
+  }
+
+  update(collection, entities, options) {
+    return this._processBatch(collection, RequestMethod.PUT, entities, options);
+  }
+
+  deleteById(collection, entityId, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.DELETE, null, null, entityId, options);
+    return this._makeHttpRequest(requestConfig);
+  }
+
+  delete(collection, query, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.DELETE, null, query, null, options);
+    return this._makeHttpRequest(requestConfig);
+  }
+
+  count(collection, query, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.GET, null, query, '_count', options);
+    return this._makeHttpRequest(requestConfig);
+  }
+
+  _processBatch(collection, method, entities, options) {
+    const isSingle = !Array.isArray(entities);
+    const requestPromises = ensureArray(entities).map((entity) => {
+      const id = method === RequestMethod.PUT ? entity._id : null; // TODO: this isn't great :)
+      const requestConfig = this._buildRequestConfig(collection, method, entity, null, id, options);
+      return this._makeHttpRequest(requestConfig);
+    });
+    // TODO: different error handling?
+    return Promise.all(requestPromises)
```

Comment:
There are no restrictions. If someone wants to make 1000 requests, they should be allowed to, imo. It's their problem how much memory their app is allowed to use, or is available, or how the platform manages http connections or a number of variables which are outside our control and if we introduce this sort of throttling, we are limiting the user's options without knowing enough about their setup. And it's not difficult to manage the request count on their own. I think we may introduce such throttling for something that would cause our backend to choke, if there is such a thing, but something that would handle client-side issues.. I think we can let the developer manage that.

Also, I haven't seen such a mechanism in the old design too. There was batching, but it was for push only. It did 100 at a time, I believe. If you insist, we can add it, it's easy.

Reply:
The point is not to disallow the user from doing something. On the contrary - we are exposing an API which takes an open-ended parameter (a list of items with no limits). We want to make sure the API functions under various cases of input, or fails predictably. Particularly since this is a `save()`, we have prior experience where users tried to save a very large list of items.

I understand that the previous implementation didn't handle this case for the `save()` API, but it should have. The batching on `push()` was added for the exact same reason. Since this is a net new enhancement to the SDK, I am ok with adding a separate ticket to address it - as long as we agree it is necessary.

Effectively, the question worth asking is this  - if I call a `save()` with 1k entities, what is supposed to happen? What about 10k? What is a reasonably large number that we can support by adding logic to batch requests?

Reply:
Actually, neither `create()`, nor `update()` allow for multiple inserts. They only allow for one entity to be passed.

I'm sorry, if I misunderstood. This is present in the old code and the new. I meant that the internals have no restrictions. Things like the 10k max page size don't apply offline and I've kept all repositories able to do batch operations. Validations for such things are done at the topmost level - in the data stores.

Internally, processors and such may ask repositories to do bulk create and update. Repositories do these in the optimal way for them. `NetworkRepository` does it one by one. It's done as it's told, as best as it can. Whoever is ordering the work being done should mind these considerations.

So, to sum up :)
In the past, restrictions for a per-entity `create` and `update` were in place, and this remains the case. I believe this is the same as what I said in the first post - client is allowed to make 10000 simultaneous requests. But we also expose an API which functions under various input - just the input is very limited :)
The part where we need to be careful is internals - there are no restrictions there. I think the only place where this may be a problem is during `push()` and it's handled there - kept it as it was, 100 entities at a time.

How do you like this setup?